modules/apps/portal-crypto-hash/portal-crypto-hash-test/src/testIntegration/java/com/liferay/portal/crypto/hash/processor/test/HashProcessorTest.java

```diff
@@ -0,0 +1,345 @@
+ */
+
+package com.liferay.portal.crypto.hash.processor.test;
+
+import com.liferay.arquillian.extension.junit.bridge.junit.Arquillian;
+import com.liferay.portal.crypto.hash.generation.context.HashGenerationContext;
+import com.liferay.portal.crypto.hash.generation.context.salt.SaltCommand;
+import com.liferay.portal.crypto.hash.generation.response.HashGenerationResponse;
+import com.liferay.portal.crypto.hash.processor.HashProcessor;
+import com.liferay.portal.crypto.hash.verification.context.HashVerificationContext;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Arthur Chan
+ * @author Carlos Sierra Andrés
+ */
+@RunWith(Arquillian.class)
+public class HashProcessorTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Test
+	public void testHashGenerationAndHashVerificationWithMultipleContexts()
+		throws Exception {
+
+		HashGenerationContext hashGenerationContext1 =
+			HashGenerationContext.newBuilder(
+			).pepper(
+				_PEPPER.getBytes()
+			).saltCommand(
+				SaltCommand.generateDefaultSizeSalt()
+			).hashProvider(
+				_MESSAGE_DIGEST_ALGO_1
+			);
+
+		HashGenerationResponse hashGenerationResponse1 =
+			_hashProcessor.generate(
+				_PASSWORD.getBytes(), hashGenerationContext1);
+
+		Optional<byte[]> optionalSalt1 = hashGenerationResponse1.getSalt();
+
+		HashGenerationContext hashGenerationContext2 =
+			HashGenerationContext.newBuilder(
+			).pepper(
+				_PEPPER.getBytes()
+			).saltCommand(
+				SaltCommand.generateVariableSizeSalt(_VARIABLE_SIZE)
+			).hashProvider(
+				_MESSAGE_DIGEST_ALGO_2
+			);
+
+		HashGenerationResponse hashGenerationResponse2 =
+			_hashProcessor.generate(
+				hashGenerationResponse1.getHash(), hashGenerationContext2);
+
+		Optional<byte[]> optionalSalt2 = hashGenerationResponse2.getSalt();
+
+		HashVerificationContext hashVerificationContext1 =
+			HashVerificationContext.newBuilder(
+			).pepper(
+				_PEPPER.getBytes()
+			).salt(
+				optionalSalt1.get()
+			).hashProvider(
+				_MESSAGE_DIGEST_ALGO_1
+			);
+
+		HashVerificationContext hashVerificationContext2 =
+			HashVerificationContext.newBuilder(
+			).pepper(
+				_PEPPER.getBytes()
+			).salt(
+				optionalSalt2.get()
+			).hashProvider(
+				_MESSAGE_DIGEST_ALGO_2
+			);
+
+		Assert.assertTrue(
+			_hashProcessor.verify(
+				_PASSWORD.getBytes(), hashGenerationResponse2.getHash(),
+				hashVerificationContext1, hashVerificationContext2));
+
+		Assert.assertFalse(
+			_hashProcessor.verify(
+				_WRONG_PASSWORD.getBytes(), hashGenerationResponse2.getHash(),
+				hashVerificationContext1, hashVerificationContext2));
+	}
```

Comment:
hey Arthur... I see it better now
so now I guess my next question would be if we can make `_hashProcessor.generate` return something that can be directly consumed by `_hashProcessor.verify`. We could also have `_hashProcessor.generate` accept a varargs for the `HashGenerationContexts`? 
Maybe that would mean that the `HashGenerationResponse` could include an array (or list) of HashVerificationContext (or something intermediate that we could come up with) so that it could be directly fed into the verification method. 
Does it make sense? would it be useful? because it looks like currently we can't do that because we need to extract the salt generated on each of the operations. 

I am guessing that the pepper would need to be out of this for "security", so it would always have to be fed "from outside"

do you think is it worth it to give it a try? would you like to try it if that's the case? If you work on it we can discuss tomorrow. 

Thx. 

Carlos.